builder.go

```diff
@@ -0,0 +1,247 @@
+	Build() (*Biscuit, error)
+}
+
+type builder struct {
+	rng  io.Reader
+	root sig.Keypair
+
+	symbolsStart int
+	symbols      *datalog.SymbolTable
+	facts        *datalog.FactSet
+	rules        []*datalog.Rule
+	caveats      []*datalog.Caveat
+	context      string
+}
+
+func NewBuilder(rng io.Reader, root sig.Keypair) Builder {
+	return BuilderWithSymbols(rng, root, DefaultSymbolTable)
+}
+
+func BuilderWithSymbols(rng io.Reader, root sig.Keypair, symbols *datalog.SymbolTable) Builder {
+	return &builder{
+		rng:  rng,
+		root: root,
+
+		symbolsStart: symbols.Len(),
+		symbols:      symbols,
+		facts:        new(datalog.FactSet),
+	}
+}
+
+func (b *builder) AddAuthorityFact(fact *Fact) error {
+	if len(fact.Predicate.IDs) == 0 {
+		fact.Predicate.IDs = []Atom{SymbolAuthority}
+	} else if fact.Predicate.IDs[0] != SymbolAuthority {
+		fact.Predicate.IDs[0] = SymbolAuthority
+	}
+
+	dlFact := fact.convert(b.symbols)
+	if !b.facts.Insert(dlFact) {
+		return ErrDuplicateFact
+	}
+
+	return nil
+}
+
+func (b *builder) AddAuthorityRule(rule *Rule) error {
+	if len(rule.Head.IDs) == 0 {
+		rule.Head.IDs = []Atom{SymbolAuthority}
+	} else if rule.Head.IDs[0] != SymbolAuthority {
+		rule.Head.IDs[0] = SymbolAuthority
+	}
+
+	dlRule := rule.convert(b.symbols)
+	b.rules = append(b.rules, &dlRule)
+	return nil
+}
+
+func (b *builder) AddAuthorityCaveat(rule *Rule) error {
+	b.caveats = append(b.caveats, &datalog.Caveat{Queries: []datalog.Rule{rule.convert(b.symbols)}})
+	return nil
+}
+
+func (b *builder) AddRight(resource, right string) error {
+	return b.AddAuthorityFact(&Fact{
+		Predicate: Predicate{
+			Name: "right",
+			IDs: []Atom{
+				SymbolAuthority,
+				String(resource),
+				Symbol(right),
+			},
+		},
+	})
+}
+
+func (b *builder) Build() (*Biscuit, error) {
+	return New(b.rng, b.root, b.symbols, &Block{
+		index:   0,
+		symbols: b.symbols.SplitOff(b.symbolsStart),
+		facts:   b.facts,
+		rules:   b.rules,
+		caveats: b.caveats,
+		context: b.context,
+	})
+}
+
+type Unmarshaler struct {
+	Symbols *datalog.SymbolTable
+}
+
+var DefaultUnmarshaler = &Unmarshaler{Symbols: DefaultSymbolTable}
+
+func (u *Unmarshaler) Unmarshal(serialized []byte) (*Biscuit, error) {
+	if u.Symbols == nil {
+		return nil, errors.New("biscuit: unmarshaler require a symbol table")
+	}
+
+	symbols := u.Symbols.Clone()
+
+	container := new(pb.Biscuit)
+	if err := proto.Unmarshal(serialized, container); err != nil {
+		return nil, err
+	}
+
+	pbAuthority := new(pb.Block)
+	if err := proto.Unmarshal(container.Authority, pbAuthority); err != nil {
+		return nil, err
+	}
+
+	signature, err := protoSignatureToTokenSignature(container.Signature)
+	if err != nil {
+		return nil, err
+	}
+
+	pubKeys := make([]sig.PublicKey, len(container.Keys))
+	for i, pk := range container.Keys {
+		pubKey, err := sig.NewPublicKey(pk)
+		if err != nil {
+			return nil, err
+		}
+		pubKeys[i] = pubKey
+	}
+
+	signedBlocks := [][]byte{container.Authority}
```

Comment:
You should allocate this as a slice with capacity `len(container.Blocks)+1` first.